docs/rfcs/2024-08-06-json-datatype.md

```diff
@@ -0,0 +1,141 @@
+
+# Details
+
+## User Interface
+The feature introduces a new data type for the database, similar to the common JSON type. Data is written as JSON strings and can be queried using functions.
+
+For example:
+```SQL
+CREATE TABLE IF NOT EXISTS test (
+    ts TIMESTAMP TIME INDEX,
+    a INT,
+    b JSON
+);
+
+INSERT INTO test VALUES(
+    0,
+    0,
+    '{
+        "name": "jHl2oDDnPc1i2OzlP5Y",
+        "timestamp": "2024-07-25T04:33:11.369386Z",
+        "attributes": { "event_attributes": 48.28667 }
+    }'
+);
+
+SELECT json_get(b, 'name') FROM test;
++---------------------+
+| b.name              |
++---------------------+
+| jHl2oDDnPc1i2OzlP5Y |
++---------------------+
+
+SELECT json_get_by_paths(b, 'attributes', 'event_attributes') + 1 FROM test;
++-------------------------------+
+| b.attributes.event_attributes |
++-------------------------------+
+|                      49.28667 |
++-------------------------------+
+
+```
+
+## Storage
+
+### Schema Inference
+Unlike other types, the schema of JSON data is inconsistent. For different JSON columns, we introduce a dynamic schema inference method for storing the data.
+
+For example:
+```JSON
+{
+  "a": "jHl2oDDnPc1i2OzlP5Y",
+  "b": "2024-07-25T04:33:11.369386Z",
+  "c": { "d": 48.28648 }
+}
+```
+This will be parsed at runtime and stored as a corresponding `Struct` type in Arrow:
+```Rust
+Struct(
+    Field("a", Utf8),
+    Field("b", Utf8),
+    Field("c", Struct(Field("d", Float64))),
+)
+```
+
+Dynamic schema inference helps achieve compression in some scenarios. See [benchmark](https://github.com/CookiePieWw/json-format-in-parquet-benchmark/) for more information.
+
+## Schema Change
+The schema must remain consistent for a column within a table. When inserting data with different schemas, schema changes may occur. There are two types of schema changes:
+
+1. Field Addition
+
+   Newly added fields can be incorporated into the schema, treating added fields in previously inserted data as null:
+   ```Rust
+   Struct(
+       Field("a", Utf8),
+   )
+   +
+   Struct(
+       Field("a", Utf8),
+       Field("e", Int32)
+   )
+   =
+   Struct(
+       Field("a", Utf8),
+       Field("e", Int32)
+   )
+   ```
+
+2. Field Modification
+
+   Compatible fields can be altered to the widest type, similar to integral promotion in C:
+   ```Rust
+   Struct(
+       Field("a", Int16),
+   )
+   +
+   Struct(
+       Field("a", Int32),
+   )
+   =
+   Struct(
+       Field("a", Int32),
+   )
+   ```
+
+   Non-compatible fields will fallback to a binary array to store the JSONB encoding:
+   ```Rust
+   Struct(
+       Field("a", Struct(Field("b", Float64))),
+   )
+   +
+   Struct(
+       Field("a", Int32),
+   )
+   =
+   Struct(
+       Field("a", BinaryArray), // JSONB
+   )
+   ```
+
+Like schema inference, schema changes are performed automatically without manual configuration.
```

Comment:
The current schema altering in GreptimeDB is really expensive, do we have a plan to improve it in such a case? cc @WenyXu

Reply:
I think the schema of JSON does not need to be in the table schema. This way, we only need to manage the JSON schema on the datanode without having to execute the alter table procedure. The JSON data itself carries schema information, so if you want to view the schema, directly viewing the data might also be a good way.

Reply:
> I think the schema of JSON does not need to be in the table schema. This way, we only need to manage the JSON schema on the datanode without having to execute the alter table procedure. The JSON data itself carries schema information, so if you want to view the schema, directly viewing the data might also be a good way.

I agree with you.

Reply:
Maybe we can alter the schema while building the vector in `VectorBuilder`. 
i.e.,
```rust
let mut builder = JsonBuilder::default();
builder.push(json1);
builder.push(json2); // The schema of JSON could be changed at this time.
```